core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -471,19 +472,24 @@ private static void parseObjectOrField(ParseContext context, Mapper mapper) thro
 
     private static ObjectMapper parseObject(final ParseContext context, ObjectMapper mapper, String currentFieldName) throws IOException {
         assert currentFieldName != null;
-        context.path().add(currentFieldName);
 
         ObjectMapper update = null;
         Mapper objectMapper = getMapper(mapper, currentFieldName);
         if (objectMapper != null) {
+            context.path().add(currentFieldName);
             parseObjectOrField(context, objectMapper);
+            context.path().remove();
         } else {
-            ObjectMapper.Dynamic dynamic = dynamicOrDefault(mapper, context);
+
+            // TODO: why Strings.splitStringToArray instead of String.split?
+            final String[] paths = Strings.splitStringToArray(currentFieldName, '.');
+            currentFieldName = paths[paths.length - 1];
+            Tuple<Integer, ObjectMapper> parentMapperTuple = getDynamicParentMapper(context, paths, mapper);
```

Comment:
I actually had it removing `paths.length - 1` to start with be it creates an issue when you have a inner object which has `dynamic: false` set. In this case the path added to the context will not be complete so you need to know how much of the path to remove.

I like the idea of doing to adding of the path to the context outside of the `getDynamicParentMapper` method, this should solve the above too since we will know if we successfully created the parentMapper and therefore whether to add (and subsequently remove) the path from the context. It will however mean that `getDynamicParentMapper` will need to return null if the parent mapper is not created because it contains an object with `dynamic:false`.